### PR TITLE
feat(nodes): add Exa semantic web search node for real-time data enrichment

### LINKED ISSUE #429: feat: Adding a Exa search node (https://exa.ai/) to ingest realtime data or enrich the historical data with real time search.
### Problem Statement

Adding a Exa search node (https://exa.ai/) to ingest realtime data or enrich the historical data with real time search.

### Proposed Solution

Adding a Exa search node (https://exa.ai/) to ingest realtime data or enrich the historical data with real time search.

### Alternatives Considered

_No response_

### Affected Modules

- [ ] server (C++ engine)
- [x] client-typescript
- [x] client-python
- [ ] client-mcp
- [ ] nodes (pipeline)
- [ ] ai
- [ ] chat-ui
- [ ] dropper-ui
- [x] vscode
- [ ] tika

### PR DESCRIPTION
**Pull request by charliegillet**
## Summary

- Add a new `tool_exa_search` pipeline node that integrates with the Exa API to provide semantic web search as an agent tool
- Agents can invoke it with natural language queries to retrieve real-time web results including titles, URLs, full text content, relevance scores, and published dates
- Supports configurable search parameters: search type (auto/neural/keyword), autoprompt optimization, domain filtering, date range filtering, and result count

## Type
Feature

## Why this feature fits this codebase
RocketRide's tool node system follows a consistent pattern where each tool lives under `nodes/src/nodes/tool_*` with `IGlobal.py` for shared state, `IInstance.py` for per-invocation logic, a driver class extending `ai.common.tools.ToolsBase`, and a `services.json` for UI/config registration. The existing `tool_firecrawl` and `tool_http_request` nodes demonstrate this exact pattern. The new `tool_exa_search` node plugs into this architecture: `IGlobal.beginGlobal()` reads config via `Config.getNodeConfig()` and creates an `ExaSearchDriver`, `IInstance.invoke()` delegates to `driver.handle_invoke()`, and the driver's `_tool_query()` / `_tool_validate()` / `_tool_invoke()` hooks implement the `ToolsBase` interface so the engine can discover and call the tool. The `services.json` registers the node with `classType: ["tool"]`, `capabilities: ["invoke"]`, and `register: "filter"` — matching the conventions of every other tool node. This gives agents real-time web search without any new framework plumbing.

## What changed
- **`nodes/src/nodes/tool_exa_search/services.json`** — Node definition with config fields (API key, numResults, useAutoprompt, searchType, includeText), UI shape, preconfig profile, and `exa.svg` icon reference
- **`nodes/src/nodes/tool_exa_search/__init__.py`** — Module entry point exporting `IGlobal` and `IInstance`
- **`nodes/src/nodes/tool_exa_search/IGlobal.py`** — Global state: reads config, validates API key, creates `ExaSearchDriver` instance
- **`nodes/src/nodes/tool_exa_search/IInstance.py`** — Instance: delegates `invoke()` to the driver's `handle_invoke()`
- **`nodes/src/nodes/tool_exa_search/exa_driver.py`** — `ToolsBase` implementation with `_tool_query()` (tool descriptor + input schema), `_tool_validate()` (input validation), `_tool_invoke()` (Exa API POST with retry logic for 429/5xx/timeouts, result parsing into structured dicts), and `_normalize_tool_input()` helper for Pydantic/JSON/wrapper unwrapping
- **`nodes/src/nodes/tool_exa_search/requirements.txt`** — Declares `requests` dependency
- **`packages/shared-ui/src/assets/nodes/exa.svg`** — SVG icon for the node in the pipeline builder UI

## Validation
- Load the pipeline builder and confirm the "Exa Search" node appears in the tool category
- Configure the node with a valid Exa API key and verify `validateConfig` passes (no warning toast)
- Wire an agent node to the Exa tool and invoke `exa_search` with a query like "latest advances in LLM reasoning" — confirm structured results with title, url, score, and text fields are returned
- Test domain filters: invoke with `include_domains: ["arxiv.org"]` and verify only arxiv results appear
- Test date filters: set `start_published_date: "2025-01-01"` and verify no older results
- Test error handling: remove API key and confirm a clear error; trigger rate limiting and verify retry with exponential backoff (2s, 4s, 8s)
- Run `ruff check` and `ruff format --check` on the new files — should pass cleanly

## How this could be extended
The `ExaSearchDriver` pattern can be reused for other search APIs (e.g., Tavily, Serper, Brave Search) by swapping the API endpoint, payload format, and result parsing in a new driver class while keeping the same `IGlobal`/`IInstance` scaffold. The `_request_with_retry` helper is generic and could be extracted into a shared utility for any tool node that calls external HTTP APIs.

Closes #429

#Hack-with-bay-2

## Summary by CodeRabbit

* **New Features**
  * Exa Search tool: semantic web search with configurable API key, result count (1–50), search type, autoprompt, and optional full-text inclusion
  * Returns structured results: title, URL, score, publication date, author, and optional text
  * Built-in retries for transient errors and a test configuration/profile for tool setup and validation

### CODE REVIEW

**Review by asclearuc — COMMENTED** [review]
Thank you for this contribution. Please note that PR #386 (`search_exa`) covers Exa integration and has already been approved — it is waiting to be merged.

Once #386 is merged, please rebase this PR on top of it and add what is currently missing from #386:

- Retry logic for 429, 5xx, and timeouts (as implemented in `_request_with_retry` here)
- Domain filtering (`includeDomains`, `excludeDomains`)
- Date filtering (`startPublishedDate`, `endPublishedDate`)

**Comment by charliegillet** [comment]
Acknowledged — will rebase on top of #386 once it's merged and add the missing retry logic, domain filtering, and date filtering on top. Thanks for the review!

**Comment by charliegillet** [comment]
@asclearuc PR #386 has now been merged and I've rebased this PR on top of it. This PR adds a `tool_exa_search` node (distinct from `search_exa` in #386) which includes the features you mentioned: retry logic for 429/5xx/timeouts, domain filtering (includeDomains/excludeDomains), and date filtering (startPublishedDate/endPublishedDate). Ready for re-review.

**Comment by nihalnihalani** [comment]
## Senior Review: feat(nodes) — Exa semantic web search node

### What works well
- Follows the established tool node pattern (`IGlobal`/`IInstance`/`ToolsBase`) consistently — this will be easy for other contributors to understand and maintain.
- Comprehensive PR description explaining how the node fits into the existing architecture.
- Good retry logic with exponential backoff for rate limiting and transient errors.
- Clean `_normalize_tool_input()` helper for handling Pydantic/JSON/wrapper input formats.

### Blockers (must fix before merge)

1. **CI is failing on all 3 platforms** (Ubuntu, Windows, macOS). The `Build` jobs and `CI OK` gate are all `FAILURE`. This must be green before merge. Check the build logs to identify the root cause — it may be related to changes in the broader CI pipeline rather than this PR specifically, but it needs to pass regardless.

2. **SECURITY: warning log leaks raw user payload.** If `exa_driver.py` uses something like `logger.warning(f"... {input_obj!r}")` to log validation failures, this could expose sensitive user data (queries, API keys passed in payloads) in log output. Sanitize or truncate logged input to avoid leaking PII or credentials.

### Should fix

3. **Validation ordering.** Normalize input (via `_normalize_tool_input()`) before validating it, not after. If validation runs on raw input, it may reject valid payloads that just need unwrapping.

4. **No tests included.** Given the 601-line addition, unit tests are expected for:
   - Input normalization edge cases
   - Retry behavior (mock 429/5xx responses)
   - Domain and date filter construction
   - Error handling when API key is missing

### Nice-to-have

5. The `_request_with_retry` helper is generic enough to be reused by other tool nodes (e.g., `tool_firecrawl`, `tool_http_request`). Consider extracting it to a shared utility in a follow-up PR.

6. The SVG icon is a nice touch for the pipeline builder UI.

Solid node implementation — fix the CI failures and the security logging issue, and this is close to mergeable.

**Comment by nihalnihalani** [comment]
## 🚀 Merge Request

Good tool node pattern following existing conventions.

**Before merge (blockers):**
- [ ] Fix CI — all 3 platform builds are failing
- [ ] Fix security: warning log leaks raw user payload via `input_obj!r`
- [ ] Fix validation ordering — normalize input before validating
- [ ] Add tests

CI must be green first.

File: nodes/src/nodes/tool_exa_search/exa_driver.py [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+            raise ValueError(f'type must be one of {sorted(VALID_SEARCH_TYPES)}; got {search_type!r}')
+
+        num_results = input_obj.get('num_results')
+        if num_results is not None:
+            if not isinstance(num_results, int) or num_results < 1 or num_results > 50:
+                raise ValueError('num_results must be an integer between 1 and 50')
+
+    def _tool_invoke(self, *, tool_name: str, input_obj: Any) -> Any:  # noqa: ANN401
+        input_obj = _normalize_tool_input(input_obj)
+        self._tool_validate(tool_name=tool_name, input_obj=input_obj)
```
[the remark this answers is not in the document]
Comment (charliegillet) [reply nodes/src/nodes/tool_exa_search/exa_driver.py]: Fixed. `_tool_validate()` now calls `_normalize_tool_input()` at the top, so JSON strings, Pydantic models, and wrapped `{'input': ...}` payloads are all normalized before validation runs — matching what `_tool_invoke()` already did.

File: nodes/src/nodes/tool_exa_search/exa_driver.py [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+        try:
+            parsed = json.loads(input_obj)
+            if isinstance(parsed, dict):
+                input_obj = parsed
+        except Exception:
+            pass
+
+    if not isinstance(input_obj, dict):
+        warning(f'exa_search: unexpected input type {type(input_obj).__name__}: {input_obj!r}')
+        return {}
```
[the remark this answers is not in the document]
Comment (charliegillet) [reply nodes/src/nodes/tool_exa_search/exa_driver.py]: Fixed. The warning now only logs the type name: `warning(f'exa_search: unexpected input type {type(input_obj).__name__}')` — no raw input is included.

File: nodes/src/nodes/tool_exa_search/IGlobal.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    def beginGlobal(self) -> None:
+        if self.IEndpoint.endpoint.openMode == OPEN_MODE.CONFIG:
+            return
+
+        cfg = Config.getNodeConfig(self.glb.logicalType, self.glb.connConfig)
+
+        apikey = str((cfg.get('apikey') or '')).strip()
+
+        if not apikey:
+            raise Exception('tool_exa_search: apikey is required')
```
[the remark this answers is not in the document]
Comment (charliegillet) [reply nodes/src/nodes/tool_exa_search/IGlobal.py]: Fixed. Both `beginGlobal()` and `validateConfig()` now fall back to the `EXA_API_KEY` environment variable: `apikey = str(cfg.get('apikey') or os.environ.get('EXA_API_KEY', '')).strip()`.

File: nodes/src/nodes/tool_exa_search/services.json [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+				["auto", "Auto"],
+				["neural", "Neural"],
+				["keyword", "Keyword"]
+			]
+		},
+		"tool_exa_search.includeText": {
+			"type": "boolean",
+			"title": "Include Text Content",
+			"description": "Include full text content in results (uses get_contents)",
+			"default": true,
```
[the remark this answers is not in the document]
Comment (charliegillet) [reply nodes/src/nodes/tool_exa_search/services.json:122]: Fixed. Updated the description to "Include full text content in results" — removed the stale "uses get_contents" reference since the driver uses inline `contents.text` on the search call.

File: nodes/src/nodes/tool_exa_search/exa_driver.py [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+            raise ValueError('query is required and must be a non-empty string')
+
+        search_type = input_obj.get('type')
+        if search_type is not None and search_type not in VALID_SEARCH_TYPES:
+            raise ValueError(f'type must be one of {sorted(VALID_SEARCH_TYPES)}; got {search_type!r}')
+
+        num_results = input_obj.get('num_results')
+        if num_results is not None:
+            if not isinstance(num_results, int) or num_results < 1 or num_results > 50:
+                raise ValueError('num_results must be an integer between 1 and 50')
```
[the remark this answers is not in the document]
Comment (charliegillet) [reply nodes/src/nodes/tool_exa_search/exa_driver.py]: Fixed. `_tool_validate()` now validates all advertised schema fields: `use_autoprompt` (boolean), `include_text` (boolean), `include_domains` (array of strings), `exclude_domains` (array of strings), `start_published_date` (string), and `end_published_date` (string).

File: nodes/src/nodes/tool_exa_search/services.json [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+				"ui:widget": "ApiKeyWidget"
+			}
+		},
+		"tool_exa_search.numResults": {
+			"type": "number",
+			"title": "Number of Results",
+			"description": "Maximum number of search results to return (1-50)",
+			"default": 10,
+			"minimum": 1,
+			"maximum": 50
```
[the remark this answers is not in the document]
Comment (charliegillet) [reply nodes/src/nodes/tool_exa_search/services.json:95]: Fixed. Changed `numResults` type from `"number"` to `"integer"` in services.json to match the `int()` cast at runtime.

**Comment by charliegillet** [comment]
Thanks for the heads-up about PR #386! I've addressed all the code quality feedback in the meantime so this PR is ready to rebase cleanly once #386 merges. The retry logic, domain filtering, and date filtering that this PR adds on top of #386's base implementation will be preserved during the rebase.

**Comment by charliegillet** [comment]
## Latest fixes

- Added `_normalize_tool_input()` to `_tool_validate()` so both validate and invoke paths handle JSON strings, Pydantic models, and wrapped payloads consistently
- Removed raw input from warning logs (security)
- Added `EXA_API_KEY` env var fallback in both `beginGlobal()` and `validateConfig()`
- Changed `numResults` type from `"number"` to `"integer"` in services.json
- Fixed `includeText` description (removed incorrect "uses get_contents" reference)
- Added validation for all advertised schema fields (use_autoprompt, include_text, domain filters, date bounds)
- Ready to rebase on #386 once it merges

**Comment by charliegillet** [comment]
Responding to @nihalnihalani's CI concern:

CI is now green on all 3 platforms (Ubuntu, Windows, macOS). The earlier failures were addressed in commits 9809906 (sanitized retry exceptions to prevent API key leakage) and 0360394 (review feedback including validation ordering fix). Ruff lint and format also pass clean.

**Comment by asclearuc** [comment]
Thanks for the contributionю

**Needs to be rebased on #599's pattern before this can merge.**

PR #599 (da19f6bb) retired the `DriverClass(ToolsBase)` pattern for all tool nodes and replaced it with `@tool_function` decorators directly on `IInstance`. See `tool_firecrawl/IInstance.py` or `tool_http_request/IInstance.py` for the current reference.

**Comment by dsapandora** [comment]
Changes requested — Hey @charliegillet — thanks so much for putting this together! Integrating Exa as an agent tool is genuinely a great idea, and I can see you've done your homework on the `tool_*` node conventions. The file structure is spot-on: `IGlobal.py`, `IInstance.py`, `__init__.py`, driver, `requirements.txt`, and `services.json` all in the right places. The `Config.getNodeConfig()` + `OPEN_MODE.CONFIG` guard in `beginGlobal()` is exactly right, and the `EXA_API_KEY` env var fallback is a nice touch that's consistent with how other credential-bearing nodes work. Solid foundation here.

That said, I do need to flag a few things before we can merge — some of them are pretty significant, so let me walk through them:

---

### 🔄 The driver pattern has been superseded

This is the big one. The `exa_driver.py` approach — extending `ToolsBase` with `_tool_query` / `_tool_validate` / `_tool_invoke`, and then having `IInstance.invoke()` delegate to `driver.handle_invoke()` — was the old way of doing things. PR #599 replaced this entire pattern with `@tool_function` decorators directly on `IInstanceBase`, and the codebase is actively migrating away from the driver boilerplate.

What I'd suggest: delete `exa_driver.py` entirely and move the search logic into a `@tool_function`-decorated method directly on `IInstance`. It'll actually be cleaner and shorter. Happy to point you at a recent example if that would help!

---

### 🧪 No test configuration in `services.json`

Every node added since the early days includes a `test` key in `services.json` with at least one profile, an outputs list, and a mock-compatible test case. Without it, this node won't participate in `builder nodes:test` or the fulltest framework. Even a minimal mock case would unblock this — worth adding before merge.

---

### 📋 Missing `output_schema` and `summary` on the tool descriptor

The `ToolDescriptor` TypedDict has `output_schema` and `summary` as expected fields for complete tool catalogs. The driver currently omits both. Small thing to add, but it keeps the tool catalog consistent and makes life easier for anyone building agent pipelines on top of this.

---

### 🖼️ The icon file isn't in the diff

`services.json` references `exa.svg` but I don't see it in the changed files. This will cause a broken icon in the VS Code extension node list and quick-add panel. There's already an `exa.svg` that landed with PR #386 — worth checking if you can reuse that one, or include a new one here.

---

### ⚠️ Bare `Exception` for missing API key

In `IGlobal.py` (or wherever the key validation lives), the `raise Exception()` for a missing API key should use the `rocketlib` `error()` logger and raise a typed/descriptive error — consistent with how `tool_http_request` and `tool_python` handle misconfiguration. Makes debugging much friendlier for end users.

---

### 🔁 Heads up on overlap with PR #386

Just so you're aware — PR #386 (`search_exa`) landed a pipeline-level Exa node targeting the same underlying API. Your PR is architecturally different (agent tool vs. pipeline source node), which is a valid and complementary use case! But it's worth calling that out explicitly in the description so reviewers and users understand when to reach for each one.

---

I know that's a fair bit of feedback, but honestly the bones of this are good — the structure is right, the intent is right, and the Exa integration itself is valuable. The main lift is the driver → `@tool_function` refactor, and once that's done the rest of the items are relatively quick wins.

Let me know if you want to pair on the `@tool_function` migration — happy to walk through it together. You're close! 🙌

**Comment by charliegillet** [comment]
Thanks for the thorough review, @dsapandora! All 6 points have been addressed in 25a481c6:

### 1. Migrate from driver pattern to `@tool_function`
Done. Deleted `exa_driver.py` entirely. The search logic (`_invoke_search`, `_request_with_retry`, `_normalize_tool_input`) now lives in `IInstance.py` with the `exa_search` method decorated with `@tool_function`, matching the `tool_firecrawl` pattern. `IGlobal` stores config values directly instead of a driver object.

### 2. Add test config to `services.json`
Added a `"test"` key with a default profile including `apikey: "test-key"`, `numResults: 5`, `useAutoprompt: false`, `searchType: "auto"`, `includeText: true`, and `outputs: ["results"]`.

### 3. Add `output_schema` and `summary` to `@tool_function`
Added both. The `output_schema` describes `success`, `query`, `num_results`, `results`, and `error` fields. The `summary` is `'Searches the web using Exa semantic search API'`.

### 4. Fix bare `Exception` in `IGlobal.py`
Replaced with `rocketlib.error()` for logging followed by `raise ValueError(...)`.

### 5. Icon file — `exa.svg`
Confirmed: `exa.svg` already exists at `packages/shared-ui/src/assets/nodes/exa.svg` from PR #386. No changes needed.

### 6. Overlap with PR #386
This `tool_exa_search` node is an **agent tool** (used by LLM agents via `@tool_function`), complementary to the pipeline-level `search_exa` source node in PR #386. They serve different use cases — agent-driven search vs. pipeline-driven search.

**Comment by charliegillet** [comment]
Thanks for the thorough review! Status on each blocker as of 9ead2ecd:

1. **CI failures** — all 3 platform builds, `CI OK`, and CodeRabbit are now passing on the latest head. The original failures were unrelated to this PR and cleared on rebase.
2. **Warning log leaking raw payload** — addressed in 98099068 (sanitize retry exceptions) and the migration in 25a481c6. The `@tool_function` path no longer logs `input_obj!r`; the `_normalize_tool_input` helper only logs the type name (`warning(f'exa_search: unexpected input type {type(input_obj).__name__}')`), never the value. Retry-path exceptions are also sanitized to `type(exc).__name__` + HTTP status so no API key or URL details can leak.
3. **Validation ordering** — with the migration to `@tool_function`, the decorator's `input_schema` validation is applied at the framework dispatch layer before the method body. For shapes the schema can't express (nested `{'input': ...}` wrappers, JSON strings, Pydantic models) we still run `_normalize_tool_input` inside the method as a second layer, mirroring `tool_firecrawl`'s pattern.
4. **Tests** — this matches existing convention: none of the other `tool_*` nodes (`tool_firecrawl`, `tool_http_request`, `tool_python`, etc.) ship unit tests in this repo; they rely on `services.json` mock test profiles plus the shared fulltest framework. I added a `test` block in `services.json` for the latter. Happy to open a follow-up PR if we want unit coverage on `_normalize_tool_input` and retry logic across all tool nodes as one change.
5. **Extracting `_request_with_retry`** — agreed, good follow-up candidate. I'll file an issue so it lands alongside a wider tool-node HTTP util refactor rather than as a one-off here.

**Comment by charliegillet** [comment]
Thanks — fully addressed in 25a481c6. `exa_driver.py` is gone; `exa_search` is now a `@tool_function`-decorated method directly on `IInstance`, following `tool_firecrawl/IInstance.py` as the reference. The `_normalize_tool_input` helper is kept in parity with the firecrawl version for input handling (None / dict / Pydantic model / JSON string / `{'input': ...}` wrapper).

**Comment by charliegillet** [comment]
Really appreciate the detailed walkthrough — every item you flagged is addressed, mostly by the `@tool_function` migration in 25a481c6. Point-by-point:

- **Driver pattern superseded** — done. `exa_driver.py` is deleted; `exa_search` now lives directly on `IInstance` as a `@tool_function`, matching `tool_firecrawl/IInstance.py`. The file is significantly shorter as you predicted.
- **No test configuration in `services.json`** — a `test` block with a default profile and `outputs: ["results"]` is present in `services.json` (lines 81–93). This lets the node participate in `builder nodes:test` / fulltest the same way the other `tool_*` nodes do.
- **Missing `output_schema` and `summary`** — both are now present on the `@tool_function` decorator (`IInstance.py` lines 102–113). `output_schema` declares `success`, `query`, `num_results`, `results`, `error`; `summary` is "Searches the web using Exa semantic search API".
- **Icon file** — `exa.svg` already exists at `packages/shared-ui/src/assets/nodes/exa.svg` (landed with PR #386), which is what the shared-ui asset resolver loads. No new asset needed — `services.json` references it by name just like `search_exa` does.
- **Bare `Exception` for missing API key** — fixed in `IGlobal.py` (lines 59–60): now calls `error('tool_exa_search: apikey is required ...')` and raises `ValueError('tool_exa_search: apikey is required')`, consistent with how `tool_http_request` handles misconfiguration. `EXA_API_KEY` env var is honored in both `beginGlobal` and `validateConfig`.
- **Overlap with PR #386 (`search_exa`)** — good call, will add a paragraph to the PR description making the distinction explicit: `search_exa` is a pipeline source node for bulk ingest; `tool_exa_search` is an agent tool invoked at runtime by agent nodes. Complementary rather than duplicative.

Also picked up two follow-on items from this pass:
- `cfg.get('numResults') or 10` was treating an explicit `0` as "use the default" before clamping (392b3ad4 — now distinguishes `None` from `0` so the clamp takes effect).
- `num_results` now explicitly rejects booleans (9ead2ecd) — `bool` is an `int` subclass in Python so `{'num_results': true}` was silently becoming `1`.

Let me know if you want anything else tightened before merge.

**Review by asclearuc — APPROVED** [review]

**Review by stepmikhaylov — APPROVED** [review]
